File: digitalMenu/backend/src/main/java/com/digitalmenu/backend/business/service/UserService.java
package com.digitalmenu.backend.business.service;

import com.digitalmenu.backend.business.DTOs.usersDTOS.LoginRequestDTO;
import com.digitalmenu.backend.business.DTOs.usersDTOS.UserRequestDTO;
import com.digitalmenu.backend.business.DTOs.usersDTOS.UserResponseDTO;
import com.digitalmenu.backend.business.mapper.UserMapper;
import com.digitalmenu.backend.business.model.User;
import com.digitalmenu.backend.infrastructure.exception.custom.EmailAlreadyExistsException;
import com.digitalmenu.backend.infrastructure.exception.custom.InvalidCredentialsException;
import com.digitalmenu.backend.infrastructure.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class UserService {
    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final UserMapper userMapper;

    public UserResponseDTO register(UserRequestDTO dto) {
        //verifica se ja existe um usuário com o email informado
        if (userRepository.findByEmail(dto.getEmail()).isPresent()) {
            throw new EmailAlreadyExistsException("Email ja cadastrado.");
        }
        //cria entidade a partir do dto
        User user = userMapper.toEntity(dto);

        //criptografa a senha
        user.setPassword(passwordEncoder.encode(user.getPassword()));

        //salva user
        User savedUser = userRepository.save(user);

        return userMapper.toResponseDTO(savedUser);
    }

    public UserResponseDTO login(LoginRequestDTO login) {
        //busca user pelo email
        User user = userRepository.findByEmail(login.getEmail())
                .orElseThrow(() -> new InvalidCredentialsException("E-mail ou senha inválidos."));
        //verigica se a senha é válida
        if(!passwordEncoder.matches(login.getPassword(), user.getPassword())) {
            throw new InvalidCredentialsException("E-mail ou senha inválidos.");
        }

        return userMapper.toResponseDTO(user);

    }
}

File: digitalMenu/backend/src/main/kotlin/com/josefilipe/digitalmenu/backend/service/UserService.kt
package com.josefilipe.digitalmenu.backend.service

import com.josefilipe.digitalmenu.backend.dto.UserRequestDTO
import com.josefilipe.digitalmenu.backend.dto.UserResponseDTO
import com.josefilipe.digitalmenu.backend.model.User
import com.josefilipe.digitalmenu.backend.repository.UserRepository
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service

@Service
class UserService (
    private val userRepository: UserRepository,
    private val passwordEncoder: PasswordEncoder
){
        //function to create user
    fun create(userRequest: UserRequestDTO): UserResponseDTO {
        val user = User(
            name = userRequest.name,
            email = userRequest.email,
            password = passwordEncoder.encode(userRequest.password),
            type = userRequest.type
        )
        val savedUser = userRepository.save(user)

        return UserResponseDTO(
            id = savedUser.id,
            name = savedUser.name,
            email = savedUser.email,
            type = savedUser.type
        )
    }

}